logging/src/logging.cpp

```diff
@@ -0,0 +1,432 @@
+    public io::basic_slot<io::base_log::get>
+{
+public:
+    typedef io::base_log::get event_type;
+    typedef typename io::basic_slot<event_type>::dispatch_type dispatch_type;
+    typedef typename io::basic_slot<event_type>::tuple_type tuple_type;
+    // This one is type of upstream.
+    typedef typename io::basic_slot<event_type>::upstream_type upstream_type;
+
+    logging_slot_t(logging_t& _parent) :
+        parent(_parent)
+    { }
+
+    virtual
+    boost::optional<std::shared_ptr<const dispatch_type>>
+    operator()(tuple_type&& args, upstream_type&&) {
+        return tuple::invoke(std::forward<tuple_type>(args), [&](std::string name) {
+            auto metafilter = parent.impl->metafilters.apply(
+                [&](std::map<std::string, std::shared_ptr<logging::metafilter_t>>& metafilters) {
+                    auto& mf = metafilters[name];
+                    if(mf == nullptr) {
+                        std::unique_ptr<logging::logger_t> internal_logger(new blackhole::wrapper_t(*(parent.impl->internal_logger), {{"metafilter", name}}));
+                        mf = std::make_shared<logging::metafilter_t>(std::move(internal_logger));
+                    }
+                    return mf;
+                }
+            );
+            std::shared_ptr<const dispatch_type> dispatch = std::make_shared<const named_logging_t>(parent.impl->logger, std::move(name), std::move(metafilter));
+            return boost::make_optional(std::move(dispatch));
+        });
+    }
+
+private:
+    logging_t& parent;
+};
+
+struct dynamic_visitor_t {
+    void
+    operator()(const dynamic_t::array_t&) {
+        throw std::system_error(std::make_error_code(std::errc::invalid_argument), "can not process array as attribute value");
+    }
+
+    void
+    operator()(const dynamic_t::object_t&) {
+        throw std::system_error(std::make_error_code(std::errc::invalid_argument), "can not process map as attribute value");
+    }
+
+    void
+    operator()(dynamic_t::null_t) {
+        throw std::system_error(std::make_error_code(std::errc::invalid_argument), "can not process null as attribute value");
+    }
+
+    template<class T>
+    void
+    operator()(const T& value) {
+        attributes.emplace_back(blackhole::attribute_t(name, value));
+    }
+
+    blackhole::attributes_t& attributes;
+    const std::string& name;
+    typedef void result_type;
+};
+
+void
+logging_t::deleter_t::operator()(impl_t* ptr) {
+    delete ptr;
+};
+
+const io::basic_dispatch_t&
+logging_t::prototype() const {
+    return *this;
+}
+
+logging_t::logging_t(context_t& context, asio::io_service& asio, const std::string& service_name, const dynamic_t& service_args) :
+    service_t(context, asio, service_name, service_args),
+    dispatch<io::base_log_tag>(service_name),
+    impl(new impl_t(
+        context,
+        [&]() {
+            auto registry = blackhole::registry_t::configured();
+            registry.add<blackhole::sink::console_t>();
```

Comment:
JFYI, the created repository allows only to create string -> console handlers. Maybe I should include all available formatters and sinks into configured registry myself in Blackhole?